src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotUseInsecureDeserializerBinaryFormatterWithoutBinderTests.cs

```diff
@@ -0,0 +1,602 @@
+            }
+        }
+    }
+}",
+                GetCSharpResultAt(14, 24, BinderNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+
+        }
+
+        [Fact]
+        public void TwoSettersNoBinder_Diagnostic()
+        {
+            VerifyCSharpWithMyBinderDefined(@"
+using System;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public object BfUnsafeDeserialize(byte[] bytes1, byte[] bytes2)
+        {
+            if (Environment.GetEnvironmentVariable(""USEFIRST"") == ""1"")
+            {
+                return new BinaryFormatter().Deserialize(new MemoryStream(bytes1));
+            }
+            else
+            {
+                return new BinaryFormatter().Deserialize(new MemoryStream(bytes2));
+            }
+        }
+    }
+}",
+                GetCSharpResultAt(14, 24, BinderNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"),
+                GetCSharpResultAt(18, 24, BinderNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+
+        }
+
+        [Fact]
+        public void BinderSetInline_NoDiagnostic()
+        {
+            VerifyCSharpWithMyBinderDefined(@"
+using System;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public object BfDeserialize(byte[] bytes)
+        {
+            return (new BinaryFormatter() { Binder = new MyBinder() }).Deserialize(new MemoryStream(bytes));
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public void Serialize_NoDiagnostic()
+        {
+            VerifyCSharp(@"
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public byte[] S(object o)
+        {
+            BinaryFormatter formatter = new BinaryFormatter();
+            MemoryStream ms = new MemoryStream();
+            formatter.Serialize(ms, o);
+            return ms.ToArray();
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public void Deserialize_LoopBinderSetAfter_Diagnostic()
+        {
+            VerifyCSharpWithMyBinderDefined(@"
+using System;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public void D(byte[][] bytes, object[] objects)
+        {
+            BinaryFormatter formatter = new BinaryFormatter();
+            for (int i = 0; i < 2; i++)
+            {
+                objects[i] = formatter.Deserialize(new MemoryStream(bytes[i]));
+                formatter.Binder = new MyBinder();
+            }
+        }
+    }
+}",
+                GetCSharpResultAt(15, 30, BinderNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
```

Comment:
>BinderNotSetRule [](start = 42, length = 16)

I believe this would be MayBeSetRule for a defensive two pass approach, but this looks fine.

Reply:
Aye, I want the more aggressive approach, since the first iteration is definitely vulnerable code.